refactor cert-gen logic

dungdm93 opened the pull request:
This is third part of refactoring helm chart #16792 aim to refactor cert-gen logic.

Currently, the cilium chart support 3 different types of cert-gen: provided, helm-gen, cronjob-gen. But the logic between those 3 types is not the same, for example: helm-gen method always generate both CA and certs (ignore user provided), but cronjob-gen reuse and disable generate cert when user provide CA/cert (this mean no cert auto-rotate).
https://github.com/cilium/cilium/blob/703b38fd47da1f91a342a8546f53542d624d8b22/install/kubernetes/cilium/templates/_hubble-generate-certs-job-spec.tpl#L47-L53  
So I simplify and unify this logic as following:
* Split each tls method to its own folder:
![Screenshot from 2021-07-02 10-45-30](https://user-images.githubusercontent.com/6848311/124217492-a6476580-db22-11eb-9da4-31055cd0efc9.png)
* When certificate is provided (`not auto.enabled`), both `cert` and `key` are required
* When certificate `auto.enabled` (both helm and cronjob methods), if user provide CA, use it; else generate new one. Other client/server/remote TLS are generated.

Signed-off-by: Đặng Minh Dũng <dungdm93@live.com>

dungdm93 committed:
refactor cert-gen logic

Signed-off-by: Đặng Minh Dũng <dungdm93@live.com>

errordeveloper commented on install/kubernetes/cilium/templates/_helpers.tpl:
```diff
@@ -5,6 +5,14 @@ Create chart name and version as used by the chart label.
 {{- end }}
 
+{{/*
+Return fqdn cilium image
```
I don't understand this comment, can you please take a look and see if you can rephrase it? Seems like the reference to FQDN could be misused?

errordeveloper reviewed (commented):
I don't know this enough to review it, let's try to ask someone who authored this part of the Helm chart. I think @kaworu or @gandro should be able to help :)

gandro commented on install/kubernetes/cilium/templates/_helpers.tpl:
```diff
@@ -33,105 +41,3 @@ backend:
-
-
-{{/*
-Generate TLS certificates for Hubble Server and Hubble Relay.
-
-Note: these 2 lines, that are repeated several times below, are a trick to
-ensure the CA certs are generated only once:
-
-    $ca := .ca | default (genCA "hubble-ca.cilium.io" (.Values.hubble.tls.auto.certValidityDuration | int))
-    $_ := set . "ca" $ca
-
-Please, don't try to "simplify" them as without this trick, every generated
-certificate would be signed by a different CA.
```
Please retain this comment where ever the `$_ := set . "ca" $ca` pattern is used.

dungdm93 replied:
This function is move to `templates/hubble/tls-helm/_helpers.tpl` because it's helm certgen specific.
I also keep the comment back, but change a little bit to match with new context. Could you please take a look @gandro

gandro replied:
Thank you! I proposed a bit a more in-depth adjustment below.

gandro commented on install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/cronjob.yaml:
```diff
@@ -3,7 +3,6 @@ apiVersion: batch/v1beta1
 kind: CronJob
 metadata:
   name: clustermesh-apiserver-generate-certs
-  namespace: {{ .Release.Namespace }}
```
This must not be removed. This is needed for `helm template -n NAMESPACE` to work correctly. See https://github.com/helm/helm/issues/3553#issuecomment-521424132

I locally tested this branch with helm 3.6.3 running `helm template --namespace kube-system` and the generated YAML was indeed missing the namespace in the cronjob.

**Note**: This is not the only place where you removed the namespace. To avoid unnecessary noise, I'll just point it out here and in the new files, but it applies to any resource touched by this PR.

dungdm93 replied:
`namespace` is back. Could you please take a look

gandro commented on install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/job.yaml:
```diff
@@ -0,0 +1,19 @@
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: clustermesh-apiserver-generate-certs-{{$checkSum}}
```
This (and all other resources moved to a separate file) needs an explicit `namespace: {{ .Release.Namespace }}` as per the comment above.

gandro replied:
But thanks for adding the checksum to this name here, btw!

dungdm93 replied:
`namespace` is back

gandro commented on install/kubernetes/cilium/templates/clustermesh-apiserver/tls-cronjob/_job-spec.tpl:
```diff
@@ -23,25 +21,23 @@ spec:
-            {{- end }}
-            {{- if not (and .Values.clustermesh.apiserver.tls.client.cert .Values.clustermesh.apiserver.tls.client.key) }}
+            {{- if .Values.externalWorkloads.enabled }}
             - "--clustermesh-apiserver-client-cert-generate"
```
I did a quick diff, and this flag is absent in the default output in the new version. I'm not familiar enough with external workloads, so it is possible that the previous version was broken, but I wonder the change is intended or not. I can take a closer look once the other issues are addressed.

dungdm93 replied:
AFAIU, `client-cert` is used by externalWorkload only. 
So I make this change to align this with tls-provided and tls-helm certgen methods:
https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/clustermesh-apiserver/client-secret.yaml#L1

gandro commented on install/kubernetes/cilium/templates/hubble/tls-helm/ca-secret.yaml:
```diff
@@ -0,0 +1,11 @@
+{{- if and .Values.hubble.enabled .Values.hubble.tls.enabled .Values.hubble.tls.auto.enabled (eq .Values.hubble.tls.auto.method "helm") }}
```
This needs a check for `and (.Values.agent (not .Values.preflight.enabled))`. Otherwise, these secrets are created in the preflight check, which is wrong (as all other values are true here by default).

https://docs.cilium.io/en/v1.10/operations/upgrade/#running-pre-flight-check-required

You removed the `and (.Values.agent (not .Values.preflight.enabled))` check everywhere. I'm a bit nervous about that, we had issues in the past where the preflight check created artifacts that later broke the installation.

dungdm93 replied:
Currently `and (.Values.agent (not .Values.preflight.enabled))` check is only apply to `tls-provided` and `tls-helm`.  
`tls-cronjob` is not using this check:
https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble/generate-certs-cronjob.yaml#L1
And deployment is not using as well:
https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble-relay/deployment.yaml#L1
Never the less, not all certs of `tls-provided` and `tls-helm` are using this check, named [`hubble-relay-client-certs`](https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble/relay-tls-client-secret.yaml#L1), [`hubble-relay-server-certs`](https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble/relay-tls-server-secret.yaml#L1) and [`hubble-ui-client-certs`](https://github.com/cilium/cilium/blob/901afa5131f27406c43a508356469141b6ed1ce0/install/kubernetes/cilium/templates/hubble/ui-tls-client-secret.yaml#L1)

So, my point is this check is not universal, my opinion this remove this and let `hubble.enabled` take care the rest.
WDYT, @gandro?

gandro replied:
> So, my point is this check is not universal, my opinion this remove this and let hubble.enabled take care the rest.
WDYT, @gandro?

The problem is that with the values we have in the docs, https://docs.cilium.io/en/v1.10/operations/upgrade/#running-pre-flight-check-required, users will have `hubble.enabled=true`. So just checking for `hubble.enabled` will not work with the existing documentation.

If users follow the documentation as it is right now, to run the preflight check, they will have the following values:

```ini
preflight.enabled=true # explicitly set by the user
agent=false # explicitly set by the user
operator.enabled=false  # explicitly set by the user
hubble.enabled=true # default value
hubble.tls.auto.enabled=true # default value
hubble.tls.auto.method="helm" # default value
hubble.relay.enabled=false # default value
```

This is why the `tls-helm` needs to check for `preflight.enabled` and `agent`, because all other values are default to true, so it will be applied by default (even though it shouldn't).

**Basically the rule is: If a component is enabled by default, then it needs to check for `preflight.enabled` to disable itself in preflight mode.** This is why `hubble-relay` or `tls-cronjob` currently do not have the preflight check, because they are not enabled by default.

But I fully agree that the inconsistency currently is not great. So let's fix it properly and as simple as possible.
My proposal would be the following.

- `hubble-server-certs` should only be created if `agent` is true.
   - Because it's only used by cilium-agent.
  -  In both modes: `tls-helm` and `tls-provided`.
  - The full check would be: `and .Values.agent .Values.hubble.enabled .Values.hubble.tls.enabled`
- `hubble-ca-secret` should only be created if `.agent` or  `hubble.relay.enabled` are true
  -  Because it's used by Agent or Relay, or UI (but UI requires Relay, so checking only for Relay is fine)
  -  In all three modes:`tls-helm`, `tls-provided`, `tls-cronjob`
  - The full check would be: `and (or .Values.agent .Values.hubble.relay.enabled) .Values.hubble.enabled .Values.hubble.tls.enabled`

The rest (`hubble-relay-server-certs`, `hubble-relay-client-certs`, `hubble-ui-client-certs`) already have to correct checks, they check for `relay` (which is disabled by default).

dungdm93 replied:
Thank for your suggestion, it's updated.
I also checked `clustermesh` tls to match the logic, but I think you need to double check.

gandro commented on install/kubernetes/cilium/templates/hubble/tls-helm/relay-server-secret.yaml:
```diff
@@ -0,0 +1,15 @@
+{{- if and .Values.hubble.enabled .Values.hubble.tls.enabled .Values.hubble.tls.auto.enabled (eq .Values.hubble.tls.auto.method "helm") .Values.hubble.relay.enabled .Values.hubble.relay.tls.server.enabled }}
```
Same for this, this is created by default in the `preflight` check. We need to check for `and (.Values.agent (not .Values.preflight.enabled))` here.

gandro reviewed (changes requested):
Thank you! I did a first pass based on what I spotted by running a few local diffs of `helm template` with different values.

gandro commented on install/kubernetes/cilium/templates/hubble/tls-helm/_helpers.tpl:
```diff
@@ -0,0 +1,24 @@
+    {{- $_ := include "hubble-generate-certs.helm.setup-ca" . -}}
+
+Please, don't try to "simplify" them as without this trick, every generated
+certificate would be signed by a different CA.
```
Thanks for adding back the comment. It needs to be adapted for the new system though:

```suggestion
Note: Always use this template as follows:

    {{- $_ := include "hubble-generate-certs.helm.setup-ca" . -}}

The assignment to `$_` is required because we store the generated CI in a global `ca` variable.
Please, don't try to "simplify" this, as without this trick, every generated
certificate would be signed by a different CA.
```

gandro commented on install/kubernetes/cilium/templates/clustermesh-apiserver/tls-helm/_helpers.tpl:
```diff
@@ -0,0 +1,24 @@
+    {{- $_ := include "clustermesh-apiserver-generate-certs.helm.setup-ca" . -}}
+
+Please, don't try to "simplify" them as without this trick, every generated
+certificate would be signed by a different CA.
```
```suggestion
Note: Always use this template as follows:

    {{- $_ := include "clustermesh-apiserver-generate-certs.helm.setup-ca" . -}}

The assignment to `$_` is required because we store the generated CI in a global `cmca` variable.
Please, don't try to "simplify" this, as without this trick, every generated
certificate would be signed by a different CA.
```

gandro reviewed (commented):
Thank you for addressing the feedback, starting to look good. We still need to resolve the secrets in the preflight check, other than that just a small adjustment regarding the comment that you added back.

gandro commented on install/kubernetes/cilium/templates/hubble/tls-cronjob/ca-secret.yaml:
```diff
@@ -0,0 +1,14 @@
+metadata:
+  name: hubble-ca-secret
+  namespace: {{ .Release.Namespace }}
+type: kubernetes.io/tls
```
This line needs to be removed, as this secret unfortunately does not conform to the `tls` type. I got the following error while testing some upgrade paths on this branch:

```
Error: UPGRADE FAILED: cannot patch "hubble-ca-secret" with kind Secret: Secret "hubble-ca-secret" is invalid: [type: Invalid value: "kubernetes.io/tls": field is immutable, data[tls.crt]: Required value, data[tls.key]: Required value]
```

dungdm93 replied:
My bad, it's updated.

gandro reviewed (changes requested):
Caught one more problem, the `kubernetes.io/tls` type cannot be used for the Hubble CA secret unfortunately.

dungdm93 committed:
tls-cronjob:

* remove deprecated `hubble-ca-cert` configmap.
  Use `hubble-ca-secret` secret instead
* no need to specify default options:
  - `*-cert-secret-name`
  - `*-cert-generate=false`
* support validity-duration option

Signed-off-by: Đặng Minh Dũng <dungdm93@live.com>

gandro reviewed (approved):
Looks good now! Thanks a lot for your work :rocket:

qmonnet commented on Documentation/operations/upgrade.rst:
```diff
@@ -324,7 +324,8 @@ Removed Options
   of ``hubble-event-buffer-capacity``. It is now removed.
-
+* The deprecated ``hubble-ca-cert`` configmap has been removed. Use
+  ``hubble-ca-secret`` secret instead.
```
Please re-add a blank line after this item and before the title of the next subsection.

dungdm93 replied:
Blank line is added

qmonnet reviewed (approved):
I'm not super familiar with the Helm chart. I had a look at the changes and nothing particularly stood out, so all good for me.
Just one nit on the Documentation part, see comment below.

qmonnet reviewed (changes requested):
[Marking as “Request changes” to leave you the time to address the nit before this gets merged]

dungdm93 committed:
Add release note to `upgrade.rst`

Signed-off-by: Đặng Minh Dũng <dungdm93@live.com>

qmonnet reviewed (approved):
Thanks!

gandro commented on install/kubernetes/cilium/templates/hubble/tls-helm/server-secret.yaml:
```diff
@@ -0,0 +1,16 @@
+{{- if and .Values.agent .Values.hubble.enabled .Values.hubble.tls.enabled .Values.hubble.tls.auto.enabled (eq .Values.hubble.tls.auto.method "helm") }}
+{{- $_ := include "hubble-generate-certs.helm.setup-ca" . -}}
+{{- $cn := list "*" (.Values.cluster.name | replace "." "-") "hubble-grpc.cilium.io" | join "." }}
+{{- $cert := genSignedCert $cn nil nil (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```
Missing SANs:
```suggestion
{{- $cert := genSignedCert $cn nil (list $cn) (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```

gandro commented on install/kubernetes/cilium/templates/hubble/tls-helm/relay-server-secret.yaml:
```diff
@@ -0,0 +1,16 @@
+{{- if and .Values.hubble.enabled .Values.hubble.tls.enabled .Values.hubble.tls.auto.enabled (eq .Values.hubble.tls.auto.method "helm") .Values.hubble.relay.enabled .Values.hubble.relay.tls.server.enabled }}
+{{- $_ := include "hubble-generate-certs.helm.setup-ca" . -}}
+{{- $cn := "*.hubble-relay.cilium.io" }}
+{{- $cert := genSignedCert $cn nil nil (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```
Missing SANs:
```suggestion
{{- $cert := genSignedCert $cn nil (list $cn) (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```

gandro commented on install/kubernetes/cilium/templates/hubble/tls-helm/relay-client-secret.yaml:
```diff
@@ -0,0 +1,16 @@
+{{- if and .Values.hubble.enabled .Values.hubble.tls.enabled .Values.hubble.tls.auto.enabled (eq .Values.hubble.tls.auto.method "helm") .Values.hubble.relay.enabled }}
+{{- $_ := include "hubble-generate-certs.helm.setup-ca" . -}}
+{{- $cn := "*.hubble-relay.cilium.io" }}
+{{- $cert := genSignedCert $cn nil nil (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```
Missing SANs:
```suggestion
{{- $cert := genSignedCert $cn nil (list $cn) (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```

dungdm93 replied:
Actually client certificate is used to authenticate to server. It's no need to specify SANs.

gandro commented on install/kubernetes/cilium/templates/hubble/tls-helm/ui-client-certs.yaml:
```diff
@@ -0,0 +1,16 @@
+{{- if and .Values.hubble.enabled .Values.hubble.tls.enabled .Values.hubble.tls.auto.enabled (eq .Values.hubble.tls.auto.method "helm") .Values.hubble.ui.enabled .Values.hubble.relay.enabled .Values.hubble.relay.tls.server.enabled }}
+{{- $_ := include "hubble-generate-certs.helm.setup-ca" . -}}
+{{- $cn := "*.hubble-ui.cilium.io" }}
+{{- $cert := genSignedCert $cn nil nil (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```
Missing SANs:
```suggestion
{{- $cert := genSignedCert $cn nil (list $cn) (.Values.hubble.tls.auto.certValidityDuration | int) .ca -}}
```

gandro reviewed (changes requested):
Thanks for the follow-ups. However, it seems like the failure in the Hubble test suite actually is legit:

```
2021-07-29T13:03:06.177107712Z level=warning msg="Failed to create gRPC client" address="192.168.36.11:4244" error="context deadline exceeded: connection error: desc = \"transport: authentication handshake failed: x509: certificate relies on legacy Common Name field, use SANs or temporarily enable Common Name matching with GODEBUG=x509ignoreCN=0\"" hubble-tls=true next-try-in=10m40s peer=k8s1 subsys=hubble-relay
```

I was able to reproduce it locally, by installing the Helm chart with Relay enabled against a local kind cluster:

```
export CILIUM_CI_TAG=$(git rev-parse HEAD)
helm upgrade --install cilium ./cilium \
  --namespace kube-system \
  --set image.repository=quay.io/cilium/cilium-ci \
  --set image.tag=$CILIUM_CI_TAG \
  --set operator.image.repository=quay.io/cilium/operator \
  --set operator.image.suffix=-ci \
  --set operator.image.tag=$CILIUM_CI_TAG \
  --set clustermesh.apiserver.image.repository=quay.io/cilium/clustermesh-apiserver-ci \
  --set clustermesh.apiserver.image.tag=$CILIUM_CI_TAG \
  --set hubble.relay.image.repository=quay.io/cilium/hubble-relay-ci \
  --set hubble.relay.image.tag=$CILIUM_CI_TAG \
  --set hubble.relay.enabled=true
```

I re-reviewed the cert sections and pointed out the missing SANs.
